stac_fastapi/types/stac_fastapi/types/rfc3339.py

```diff
@@ -45,53 +46,88 @@ def rfc3339_str_to_datetime(s: str) -> datetime:
+
+    Raises:
+        ValueError: If the date_str is empty or contains the placeholder '..'.
+    """
+    if ".." in date_str or not date_str:
+        raise ValueError("Invalid date format.")
+    return rfc3339_str_to_datetime(date_str)
+
+
+def validate_interval_format(values: list) -> None:
+    """
+    Validate the format of the interval string to ensure it contains at most
+    one forward slash.
 
     Args:
-        interval (str or None): The interval string to convert to a tuple of
-        datetime.datetime objects, or None if no datetime is specified.
+        values (list): A list of strings split by '/' from the interval string.
+
+    Raises:
+        HTTPException: If the interval string contains more than one forward slash.
+    """
+    if len(values) > 2:
+        raise HTTPException(
+            status_code=400,
+            detail="Interval string contains more than one forward slash.",
+        )
+
+
+def str_to_interval(interval: Optional[str]) -> Optional[DateTimeType]:
+    """
+    Extract a tuple of datetime objects from an interval string defined by the OGC API.
+    The interval can either be a single datetime or a range with start and end datetime.
+
+    Args:
+        interval (Optional[str]): The interval string to convert to datetime objects,
+        or None if no datetime is specified.
 
     Returns:
-        Optional[DateTimeType]: A tuple of datetime.datetime objects or None if
-        input is None.
+        Optional[DateTimeType]: A tuple of datetime.datetime objects or
+        None if input is None.
 
     Raises:
-        ValueError: If the string is not a valid interval string and not None.
+        HTTPException: If the string is not valid for various reasons such as being empty,
+        having more than one slash, or if date formats are invalid.
     """
     if interval is None:
         return None
 
     if not interval:
-        raise ValueError("Empty interval string is invalid.")
+        raise HTTPException(status_code=400, detail="Empty interval string is invalid.")
 
     values = interval.split("/")
-    if len(values) == 1:
-        # Single date for == date case
-        return rfc3339_str_to_datetime(values[0])
-    elif len(values) > 2:
-        raise ValueError(
-            f"Interval string '{interval}' contains more than one forward slash."
+    validate_interval_format(values)
+
+    try:
+        start = parse_single_date(values[0]) if values[0] not in ["..", ""] else None
+        end = (
+            parse_single_date(values[1])
+            if len(values) > 1 and values[1] not in ["..", ""]
+            else None
         )
-
-    start = None
-    end = None
-    if values[0] not in ["..", ""]:
-        start = rfc3339_str_to_datetime(values[0])
-    if values[1] not in ["..", ""]:
-        end = rfc3339_str_to_datetime(values[1])
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=str(e))
```

Comment:
I would personally raise HTTPException witting the `parse_single_date` function and remove the `try .. except` block .

Reply:
I think this is hard because the try except catches errors that filter from this function --- rfc3339_str_to_datetime via parse_single_date

Reply:
We could put the try except in parse_single_date?

Reply:
the other solution would be to change https://github.com/stac-utils/stac-fastapi/blob/73402d9b58c0291f18d3b128a1d417ccaf4bd1aa/stac_fastapi/types/stac_fastapi/types/rfc3339.py#L43C15-L43C25 and raise a specific error `InvalidDatetime` and then catch this in https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/api/stac_fastapi/api/errors.py#L23-L30

Reply:
Things seem to pass the re.match and then error out here `return iso8601.parse_date(date_str)` hence the try except: https://github.com/stac-utils/stac-fastapi/blob/73402d9b58c0291f18d3b128a1d417ccaf4bd1aa/stac_fastapi/types/stac_fastapi/types/rfc3339.py#L40-L46

Reply:
I can't remember what exactly now

Reply:
Not everything